### ISSUE TITLE
Convert hs.chooser into a non-activating NSPanel

### ISSUE THREAD

**Issue by latenitefilms** [Hammerspoon/hammerspoon#2067]
Currently `hs.chooser` uses a `NSWindow`, which doesn't support "non activating". It would be great if we could convert `hs.chooser` from a `NSWindow` to a `NSPanel` so that we can customise the `windowStyle`, like you can with a `hs.webview`. Specifically, I want to be able to activate a `hs.chooser` from another application (i.e. Safari), and when the chooser appears, Safari (not Hammerspoon) still has focus.

Discussed here: 

https://github.com/Hammerspoon/hammerspoon/pull/2062#issuecomment-479764102

**Comment by latenitefilms** [comment]
As discussed in [#2062](https://github.com/Hammerspoon/hammerspoon/pull/2062#issuecomment-479765974)...

After much messing around I finally worked out that the trick to getting `[self.window _setPreventsActivation:YES] ;` to work is that you need to disable "Hide On Deactivate" in Interface Builder. I can now make the chooser show over the top of any application, without activating Hammerspoon, but `preventsActivation` does actually what it says, and you can't actually click on the chooser, which means you can't type in it, or use the arrow keys to navigate the list (although clicking on the menu items with your mouse works).

So unfortunately, whist this undocumented API call works great for `hs._asm.undocumented.touchbar`, it won't work for `hs.chooser`, because we actually need the user to be able to interact with the chooser (i.e. type text in the text field and navigate the list).

Therefore, I think the only solution is to keep trying to convert `hs.chooser` into an `NSPanel`.

**Comment by cmsj** [comment]
I've been playing around with this a bit and I'm hitting an issue where I can't get the chooser to appear a second time - was that working for you?
I did make a few tweaks too, so maybe I broke it.

**Comment by latenitefilms** [comment]
#2083 was working fine for me - with the disclaimer that:

> @asmagill - FYI: when this code is used in conjunction with #2062, it seems adding a toolbar makes the hs.chooser non-activating (i.e. you can't actually type text into the console). I think this is due to the way you're currently applying the theWindow.styleMask in hs.webview.toolbar.attachToolbar, but I haven't found the right combination to make it work yet.

**Comment by asmagill** [comment]
@cmsj, The comments for #2062 got a bit long winded and probably should have been branched off into the separate pull sooner then they were...

@latenitefilms, this is how I see the two issues, chime in if you disagree --

There are really two requests wrapped up in this specific issue (#2067):

#2062 -- which *only* adds toolbar support to `hs.chooser`, and
#2083 -- making `hs.chooser` a non-activating panel (or at least adding the option)

They should be reviewed and considered separately.

First and foremost, #2062 does nothing concerning the non-activating panel question -- it just adds the toolbar. My example demonstrating this can be found at https://github.com/asmagill/hammerspoon-config-take2/blob/master/_scratch/chooserToolbar.lua

@cmsj, using *just* that pull (and maybe your tweaks, I'd have to see what they are), how does your testing compare with my example? Do you still see the issue where the chooser can't be shown a second time?

Once #2062 is good and landed (or changed or scrapped for something combining the two if that's really what you think should be done), then we can consider #2083, which to be honest I have not had any time to look at... I'm hoping to find some time this weekend to get back to Hammerspoon for a bit and this was one of the things I wanted to play with and see where @latenitefilms was with it.

**Comment by latenitefilms** [comment]
Spot on @asmagill. 

#2062  is basically ready to merge (I've already merged it into CommandPost), unless anyone has any objections. Was all working fine for me.

Once #2062 is merged, then we can go back and work out how to make #2083 work with these new changes.

#2083 was working for me, WITHOUT also merging in #2062 - so I don't think it'll take @asmagill too much trouble/pain to fix #2083 so that it works with the new toolbar support for `hs.chooser`. I think the issue is just the way `styleMask` are applied in `hs.webview.toolbar.attachToolbar` - I just haven't been able to find the right combination to make things work consistent both with toolbars and without toolbars.

**Comment by cmsj** [comment]
It's probably not helpful that I'm also currently working on a branch to refresh the UI of hs.chooser a little! I'm fine with @latenitefilms 's proposed approach here - merge #2062 and then figure out how to make all the non-activating stuff work.

**Comment by latenitefilms** [comment]
@cmsj & @asmagill - FYI: I've left the pull request so that the `hs.chooser` is ALWAYS non-activating, as I can't really think of a case where you'd want to make the chooser activate Hammerspoon?

**Comment by cmsj** [comment]
I've lost track of the history of this now - did we ever figure out a way that actually works for this?

**Comment by latenitefilms** [comment]
I merged the above code into CommandPost-App and it works great.

If you’re happy to make the chooser a nom-activating NSPanel, then I believe this is good to go.

**Comment by latenitefilms** [comment]
Actually, I did have to make one other tweak so that it works correctly on both Mojave & High Sierra:

https://github.com/CommandPost/CommandPost-App/commit/f29b63a31754a60baba3e506087ac822506bd1dc

**Comment by cmsj** [comment]
@latenitefilms by "above code", you mean #2904?

**Comment by latenitefilms** [comment]
@cmsj - I've created a new pull request here: #2138
